script.js

```diff
@@ -0,0 +1,156 @@
+let todoArr = [];
+let doneArr = [];
+
+/*로컬 스토리지에서 가져오기 함수*/
+const getLocalStorage = () => {
+    const localTodo = window.localStorage.getItem('todo');
+    const localDone = window.localStorage.getItem('done');
+    console.log(localTodo, localDone);
+
+    //로컬스토리지에 데이터가 있을때에만 가져온다
+    if (localTodo) {
+        todoArr = JSON.parse(localTodo);
+        todoArr.map((v) => createTodoElement(v, 'todo'));
+    }
+    if (localDone) {
+        doneArr = JSON.parse(localDone);
+        doneArr.map((v) => createTodoElement(v, 'done'));
+    }
+};
+
+//현재 배열을 로컬스토리지로 동기화
+const syncLocalStorage = () => {
+    window.localStorage.clear();
+    window.localStorage.setItem('todo', JSON.stringify(todoArr));
+    window.localStorage.setItem('done', JSON.stringify(doneArr));
+    console.log('sync');
+};
+
+/*아이템 생성하기 함수*/
+const createTodoElement = (content, isItToDo) => {
+    const todoList = document.getElementById('todo-list');
+    const doneList = document.getElementById('done-list');
+
+    const item = document.createElement('li');
+    item.className = 'item ';
+    item.className += isItToDo;
+
+    const itemContent = document.createElement('div');
+    itemContent.className = 'item-content';
+    itemContent.innerHTML = content;
+
+    const radioButton = document.createElement('div');
+    radioButton.className = 'radio-button';
+
+    const itemRemove = document.createElement('button');
+    itemRemove.className = 'item-remove';
+
+    const itemContentBox = document.createElement('div');
+    itemContentBox.className = 'item-content-box';
+
+    itemContentBox.appendChild(radioButton);
+    itemContentBox.appendChild(itemContent);
+
+    item.appendChild(itemContentBox);
+    item.appendChild(itemRemove);
+
+    if (isItToDo === 'todo') {
+        todoList.appendChild(item);
+    } else {
+        doneList.appendChild(item);
+    }
+    countTodo();
+
+    //삭제 버튼 함수
+    const removeTodo = () => {
+        if (isItToDo === 'todo') {
+            todoArr = todoArr.filter((v) => v !== content);
+        }
+        if (isItToDo === 'done') {
+            doneArr = doneArr.filter((v) => v !== content);
+        }
+        item.remove();
+        syncLocalStorage();
+        countTodo();
+        toast('삭제되었어요!');
+    };
+    itemRemove.onclick = removeTodo;
+
+    //할일 토글 함수
+    const toggleTodo = () => {
+        if (isItToDo === 'todo') {
+            todoArr = todoArr.filter((v) => v !== content);
+            doneArr.push(content);
```

Comment:
```suggestion
            doneArr = [...doneArr, content]
```
개인적으로 array 조작은 항상 assignment로 이루어지는게 좋다고 생각합니다. ES6도 최대한 활용해보자구요...!

Reply:
감사합니다!! es6를 활용하는 연습을 계속 해봐야겠습니다.

Reply:
제가 C/C++ 방식의 매서드 호출을 통한 array 조작이 익숙해서, 할당 방식이 어색하게 느껴집니다. 
array 조작이 할당 방식으로 이루어지는게 좋은 이유가 무엇인지 알려주실 수 있을까요?

동일한 맥락으로 가령 array의 특정 원소 1개를 삭제하는 과정이 필요할때 findIndex로 해당 원소의 위치를 찾고 splice로 원소를 삭제하는것 보다 filter를 사용해서 특정 조건을 만족하는 원소를 제외한 나머지 모든 원소를 배열에 재할당 하는 방법을 지향해야 할까요?

// 1번 방법
```javascript
const keyToDel = items.findIndex(<조건>);
if (keyToDel !== -1) {
    items.splice(keyToDel, 1);
}
```

// 2번 방법
```javascript
items = items.filter(<조건>);
```

Reply:
위에 쓰신 코멘트에 답변을 못달도록 되어있어서 여기에 대신 남깁니다!

js를 쓴지 오래되다 보니 1번 방식에 대해선 전혀 생각도 못했네요. 음... 다음주부터 리액트를 사용해서 상태를 관리하게 되면, 불변성을 지키면서 상태를 업데이트하게 됩니다!! 배열 자체를 수정하게 되면 리액트 안에서 상태변화를 감지하기 어려워서, 원하는대로 재렌더링이 안될수도 있습니다. 그렇기 때문에 filter로 반환된 새로운 배열을 넣어주는것으로 알고 있습니다. 불변성 유지에 대해서 찾아보시면 될것 같습니다. 바닐라JS로 사용할때는 1번방식도 좋은 방식이 아닐까요??

저도 제가 맞다고 확신은 못드립니다..ㅠㅠ 혹시 제가 틀렸다거나 더욱 자세하게 아시는 분이 있으면 코멘트 달아주시면 감사하겠습니다!!

[제가 리액트를 공부하면서 자주 봤던 사이트입니다.](https://react.vlpt.us/basic/14-array-remove.html) 도움 되실것 같아 남겨드려요.

Reply:
맞습니다! JS의 immutability(불변성) 관련된 문제인데 [이 글](https://sustainable-dev.tistory.com/156)에 자세히 나와 있어서 참고하시면 좋을 것 같습니다! 

간단히 설명드리자면, 배열의 값을 직접 변경할 경우 디버깅이 어렵고 의도치 않은 결과를 불러올 수 있기 때문에 배열의 값을 직접 변경하는 방식이 아닌 새로운 배열을 만들어 할당하는 방식을 추천합니다. 따라서 값을 직접 변경하는 `push()`, `splice()` 등의 메소드보다 새로운 배열을 반환하는 `map()`, `filter()` 혹은 `spread` 연산자를 사용하는 방식을 권장합니다. 또한 나중에 React를 사용하실 때에도 배열 불변성을 지켜 주셔야 상태 변화 감지와 성능 최적화 부분에서 용이하답니다!

Reply:
생각 못했던 부분인데 참고자료까지 친절하게 알려주시고 규진님, 주현님 너무너무 감사합니다 😭😭

Reply:
와 너무 좋은 조언들... 저도 감사히 배워갑니다